tests/integration/construct_component_failures/testcomponent/index.ts

```diff
@@ -0,0 +1,54 @@
+// Copyright 2016-2024, Pulumi Corporation.  All rights reserved.
+
+import * as pulumi from "@pulumi/pulumi";
+import * as provider from "@pulumi/pulumi/provider";
+import * as grpc from "@grpc/grpc-js";
+import * as status from "@pulumi/pulumi/proto/status_pb";
+import * as error_details from "@pulumi/pulumi/proto/error_details_pb";
+
+class Component extends pulumi.ComponentResource {
+    public readonly foo: pulumi.Output<string>;
+
+    constructor(name: string, foo: pulumi.Input<string>, opts?: pulumi.ComponentResourceOptions) {
+        super("testcomponent:index:Component", name, undefined, opts);
+
+        this.foo = pulumi.output(foo);
+
+        this.registerOutputs({
+            foo: this.foo,
+        })
+    }
+}
+
+class Provider implements provider.Provider {
+    public readonly version = "0.0.1";
+
+    async construct(name: string, type: string, inputs: pulumi.Inputs,
+              options: pulumi.ComponentResourceOptions): Promise<provider.ConstructResult> {
+        if (type != "testcomponent:index:Component") {
+            throw new Error(`unknown resource type ${type}`);
+        }
+
+        const component = new Component(name, inputs["foo"], options);
+
+	let errorDetails = new error_details.BadRequest();
+	const violation = new error_details.BadRequest.FieldViolation();
+	violation.setField("foo");
+	violation.setDescription("the failure reason");
+	errorDetails.addFieldViolations(violation);
+
+	throw new pulumi.BadRequestError(
+	    grpc.status.INVALID_ARGUMENT, "failing for a reason", errorDetails);
```

Comment:
Maybe this could be

```
throw new pulumi.BadRequestError(
    grpc.status.INVALID_ARGUMENT, "failing for a reason", [["foo", "the failure reason"]])
```

Instead?  Not sure what the most Typescripty way to do this is.